HIVE-22239

jcamachor opened the pull request.

miklosgergely commented on ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java:
```diff
@@ -935,8 +942,11 @@ else if(colTypeLowerCase.equals(serdeConstants.SMALLINT_TYPE_NAME)){
+    } else if (colTypeLowerCase.equals(serdeConstants.TIMESTAMP_TYPE_NAME)) {
+      cs.setAvgColLen(JavaDataModel.get().lengthOfTimestamp());
+      // epoch, seconds since epoch
+      cs.setRange(0, 2177452799L);
```
Is this 2038-12-31 11:59:59 PM GMT? Why is it hard coded to that? If there is a reason for that, then I suggest to introduce a more meaningful constant declaration, like this:

// Explanation why the timestamp range upper limit should be this date
private static final long TIMESTAMP_RANGE_UPPER_LIMIT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss zzz").parse("2038-12-31 23:59:59 GMT").getTime() / 1000;

jcamachor replied:
I took the corresponding seconds value for the value that had been hardcoded for `date` type. I am not sure about the original reason to choose that value but I thought that consistency was good in this case. I will add a comment.

jcamachor replied:
I do not think this is critical, but I explored a bit and this seems to be a poor choice for a heuristic, since in most cases it will lead to underestimation of the data size (since most users will not have dates starting from 1970).
I will take as lower limit `01-01-2015` (epoch for ORC) and as upper limit `12-31-2025`. Let me know if you see any cons with this approach.

kgyrtkirk commented on ql/src/test/results/clientpositive/llap/bucket_map_join_tez2.q.out:
```diff
@@ -759,50 +759,56 @@ STAGE PLANS:
+                        key expressions: _col0 (type: int)
+                        sort order: +
+                        Map-reduce partition columns: _col0 (type: int)
+                        Statistics: Num rows: 498 Data size: 1992 Basic stats: COMPLETE Column stats: COMPLETE
```
disable enhancement; seems to be interfering with the test

kgyrtkirk commented on ql/src/test/results/clientpositive/llap/subquery_select.q.out:
```diff
@@ -3918,14 +3918,14 @@ STAGE PLANS:
                       Group By Operator
                         aggregations: max(p_size)
-                        minReductionHashAggr: 0.875
+                        minReductionHashAggr: 0.0
```
the computed range is most likely empty...
these changes suggest to me that something is not entirely right....is this expected?

jcamachor replied:
Range is `15103`-`195606` for `p_partkey` column, out of 26 rows. Hence, the estimate of `1` seems right.

jcamachor commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:
```diff
@@ -316,7 +321,7 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
     protected long evaluateExpression(Statistics stats, ExprNodeDesc pred,
         AnnotateStatsProcCtx aspCtx, List<String> neededCols,
-        Operator<?> op, long currNumRows) throws SemanticException {
+        Operator<?> op, long currNumRows, boolean uniformWithinRange) throws SemanticException {
```
I have used the `AnnotateStatsProcCtx` to hold that value, thanks.

jcamachor commented on ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java:
```diff
@@ -935,8 +942,11 @@ else if(colTypeLowerCase.equals(serdeConstants.SMALLINT_TYPE_NAME)){
+    } else if (colTypeLowerCase.equals(serdeConstants.TIMESTAMP_TYPE_NAME)) {
+      cs.setAvgColLen(JavaDataModel.get().lengthOfTimestamp());
+      // epoch, seconds since epoch
+      cs.setRange(0, 2177452799L);
```
I answer to same comment to @miklosgergely above, please see my response. I used a new heuristic, I do not think existing was a very good one...

jcamachor commented on ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java:
```diff
@@ -856,8 +856,15 @@ public static ColStatistics getColStatistics(ColumnStatisticsObj cso, String tab
+          .getSecondsSinceEpoch() : null;
+      Long highVal = (csd.getTimestampStats().getHighValue() != null) ? csd.getTimestampStats().getHighValue()
+          .getSecondsSinceEpoch() : null;
+      cs.setRange(lowVal, highVal);
```
Yeah, I did not want to change the information that we store for timestamp type, note that this patch only changes the way we read the data. I agree we could use finer granularity for different types.

jcamachor commented on ql/src/test/results/clientpositive/llap/retry_failure_stat_changes.q.out:
```diff
@@ -139,25 +139,25 @@ Stage-0
+                    Select Operator [SEL_2] (rows=7/5 width=4)
                       Output:["_col0"]
-                      Filter Operator [FIL_18] (rows=1/5 width=4)
+                      Filter Operator [FIL_18] (rows=7/5 width=4)
```
If it was by-design, do you think I should disable the uniform stats for this specific test?

jcamachor commented on standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:
```diff
@@ -562,14 +562,27 @@ struct DateColumnStatsData {
 }
 
+struct Timestamp {
+1: required i64 secondsSinceEpoch
```
I do not think it is too complicated but it will imply changes in the metastore tables that store these values too. I did not want to change the internal representation of column stats for timestamp type in this patch, that is why I introduced the type but did not change the internal representation based on seconds. I created https://issues.apache.org/jira/browse/HIVE-22309.

jcamachor commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/TimestampColumnStatsAggregator.java:
```diff
@@ -0,0 +1,358 @@
+          // We have estimation, lowerbound and higherbound. We use estimation
+          // if it is between lowerbound and higherbound.
+          double densityAvg = densityAvgSum / partNames.size();
+          estimation = (long) (diff(aggregateData.getHighValue(), aggregateData.getLowValue()) / densityAvg);
```
I agree with you... Much of this logic could be simplified. There is an abstract class `ColumnStatsAggregator` but it is basically empty. We should move some of the common logic there or to a utility class. I have created another JIRA: https://issues.apache.org/jira/browse/HIVE-22310.

jcamachor commented on ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java:
```diff
@@ -856,8 +856,15 @@ public static ColStatistics getColStatistics(ColumnStatisticsObj cso, String tab
       cs.setNumNulls(csd.getBinaryStats().getNumNulls());
-    } else if (colTypeLowerCase.equals(serdeConstants.TIMESTAMP_TYPE_NAME) ||
-        colTypeLowerCase.equals(serdeConstants.TIMESTAMPLOCALTZ_TYPE_NAME)) {
+    } else if (colTypeLowerCase.equals(serdeConstants.TIMESTAMP_TYPE_NAME)) {
```
I think it is a good idea and we are not in a hurry... Let's do the right thing.
I have created https://issues.apache.org/jira/browse/HIVE-22311.

jcamachor commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/stats/annotation/StatsRulesProcFactory.java:
```diff
@@ -967,13 +979,23 @@ private long evaluateComparator(Statistics stats, AnnotateStatsProcCtx aspCtx, E
+              if (uniformWithinRange) {
+                // Assuming uniform distribution, we can use the range to calculate
+                // new estimate for the number of rows
+                return Math.round(((double) (value - minValue) / (maxValue - minValue)) * numRows);
```
Good catch. I fixed that in latest patch.

jcamachor commented:
@kgyrtkirk , I pushed a new commit containing only the range change and addressing your comments ; I will upload a follow-up for the timestamp column stats propagation in a new PR. Can you take another look?

jcamachor commented on ql/src/java/org/apache/hadoop/hive/ql/stats/StatsUtils.java:
```diff
@@ -944,7 +948,7 @@ else if(colTypeLowerCase.equals(serdeConstants.SMALLINT_TYPE_NAME)){
       cs.setAvgColLen(JavaDataModel.get().lengthOfDate());
       // epoch, days since epoch
-      cs.setRange(0, 25201);
+      cs.setRange(DATE_RANGE_LOWER_LIMIT, DATE_RANGE_UPPER_LIMIT);
```
Yeah, this is a heuristic... No matter what you do, you will always get it wrong in some cases. I guess the idea is the target the most common case. The solution to overestimation/underestimation is to compute column stats as you mentioned, we do not want to let user tune this too.
